replica-common/src/main/java/com/graphhopper/OsmHelper.java

```diff
@@ -38,27 +61,101 @@ public int getSegmentIndexForGhEdge(int ghEdgeId) {
         return ghEdgeIdToSegmentIndexMapping.getInt(pointer);
     }
 
-    public static Map<String, String> getLanesTag(long osmId, Map<Long, Map<String, String>> osmIdToLaneTags) {
-        return osmIdToLaneTags.getOrDefault(osmId, null);
+    public static Map<String, String> getWayTag(long osmId, Map<Long, Map<String, String>> osmIdToWayTags) {
+        return osmIdToWayTags.getOrDefault(osmId, null);
     }
 
-    public static String getHighwayFromOsmWay(ReaderWay way) {
-        if (way.hasTag("highway")) {
-            return way.getTag("highway");
+    public static String getTagValueFromOsmElement(ReaderElement wayOrRelation, String tagName) {
+        if (wayOrRelation.hasTag(tagName)) {
+            return wayOrRelation.getTag(tagName);
         } else {
             return null;
         }
     }
 
+    public static Map<String, String> parseWayTags(ReaderWay ghReaderWay) {
+        Map<String, String> parsedWayTagValues = Maps.newHashMap();
+
+        // Parse street name, which is a concat of `name` and `ref` tags (if present)
+        parsedWayTagValues.put(OSM_NAME_TAG, getConcatNameFromOsmElement(ghReaderWay));
+
+        // Parse highway and direction tags, plus all tags needed for determining lane counts
+        for (String wayTag : ALL_TAGS_TO_PARSE) {
+            parsedWayTagValues.put(wayTag, getTagValueFromOsmElement(ghReaderWay, wayTag));
+        }
+
+        // Remove any tags that weren't present for this Way (ie the value was parsed as null)
+        parsedWayTagValues.values().removeIf(Objects::isNull);
+        return parsedWayTagValues;
+    }
+
     // if only `name` or only `ref` tag exist, return that. if both exist, return "<ref>, <name>". else, return null
     public static String getConcatNameFromOsmElement(ReaderElement wayOrRelation) {
-        String name = null;
-        if (wayOrRelation.hasTag("name")) {
-            name = wayOrRelation.getTag("name");
-        }
+        String name = getTagValueFromOsmElement(wayOrRelation, "name");
         if (wayOrRelation.hasTag("ref")) {
             name = name == null ? wayOrRelation.getTag("ref") : wayOrRelation.getTag("ref") + ", " + name;
         }
         return name;
     }
+
+    public static void updateOsmIdToWayTags(Map<Long, Map<String, String>> osmIdToWayTags, Long osmId, Map<String, String> newTagValues) {
+        Map<String, String> currentWayTags = new HashMap<>(osmIdToWayTags.getOrDefault(osmId, Maps.newHashMap()));
+        for (String tag : newTagValues.keySet()) {
+            if (!currentWayTags.containsKey(tag)) {
+                currentWayTags.put(tag, newTagValues.get(tag));
+            }
+        }
+        osmIdToWayTags.put(osmId, currentWayTags);
+    }
+
+    // todo: can we move this logic into CustomOsmReader?
```

Comment:
nit: is this todo still relevant?

Reply:
yes, IMO - the idea here would be to integrate this custom parsing stuff (which we call separately from the initial OSM read through graphhopper does during import) into the main import/parsing code (which is in `CustomOsmReader`). This would speed things up a bit because we'd read OSM once instead of twice, but in reality might be kind of messy. The todo is kinda serving as a reminder that this is an option, even if the task itself is low priority